### ISSUE TITLE
Label Function Analysis

### ISSUE THREAD

**Issue by schopra8** [NorskRegnesentral/skweak#13]
First of all, thanks for open sourcing such an awesome project! 

Our team has been playing around skweak for a sequential labeling task, and we were wondering if there were any plans in the roadmap to include tooling that helps practitioners understand the "impact" of their label functions statistically. 

Snorkel for example, provides a [LF Analysis tool](https://snorkel.readthedocs.io/en/master/packages/_autosummary/labeling/snorkel.labeling.LFAnalysis.html) to understand how one's label functions apply to a dataset statistically (e.g., coverage, overlap, conflicts). Similar functionality would be tremendously helpful in gauging the efficacy of one's label functions for each class in a sequential labeling problem. 

Are there any plans to add such functionality down the line as a feature enhancement?

**Comment by plison** [comment]
Thanks Sahil! Yes, this would indeed be very useful. I'll definitely add it to our TODO list, but I can't promise anything in the short term though. If you're willing to implement a first version and submit a pull request, do let us know, that would be a tremendous addition to the toolkit :-)

**Comment by schopra8** [comment]
For sure! I was planning to hack something together later this week; I'll try to make it a little more robust than an overnight hack and can issue a pull request with what I build so we can see if it's substantial enough to add to add!

**Comment by plison** [comment]
Brilliant :-) Don't hesitate to ask if you have any questions!

**Comment by schopra8** [comment]
@plison Wanted to get your quick thoughts on a general approach. 

In Snorkel, each prediction tasks is assumed to be a n-way classification problem. When it comes to a function like `label_coverage` the goal is to see what fraction of the dataset is covered by any label, whatsoever. For sequence classification, it seems we'd want to have a label_coverage metric (i.e., % of dataset given a label) for each entity type. For example, if we have 3 NER classes (PER, DATE, LOC), we would want 3 metrics -- % of dataset covered for PER, % of dataset covered for DATE, and % of dataset covered for LOC. 

**Do you agree with this approach?** An alternate approach might to compute metrics for each of the NER Tags at the span level (rather than data point level). 

Under this framing we would have the following function definitions (adapted from [Snorkel's LF Analysis](https://snorkel.readthedocs.io/en/v0.9.7/packages/_autosummary/labeling/snorkel.labeling.LFAnalysis.html)). **Here I have proposed an implementation at the data-point level, rather than span level**

**Would love to get your feedback on the following definitions for a LF Analysis submodule**:
- **label_conflict()**: For each NER Tag, compute the fraction of data points with conflicting (non-abstain labels). This would occur if one label function labeled "Apple Inc." as a PER, while another label function labeled "Apple Inc." DATE.
- **label_coverage(Y=Optional[List[Doc])**: For each NER Tag, compute the fraction of data points with at least one label (e.g., 60% of data points have PER, 50% have DATE, 40% have LOC). Unlike a classification problem like Snorkel, coverage metrics aren't the most useful. For example, 60% of the data points having PER labels might actually be a good thing - if only 60% of the dataset have PER entities. Accordingly, we should likely give users the ability to provide a set of documents with labeled spans (if they have such training data) to see how much coverage is provided relative to the hand labels.
- **label_overlap()**: For each NER Tag, compute the fraction of data points with at least two (non-abstain labels). **Not sure if this function makes much sense at the data-point granularity. Realistically, this is only helpful at the span-level (e.g., 2 label functions for PER agree to mark the same span as PER)**

Still thinking through functions like`lf_empirical_accuracies` and `lf_empirical_probs` ... so I'll circle back on that -- once we get some alignment here :)

**Comment by schopra8** [comment]
Under this framing we would have the following function definitions (adapted from [Snorkel's LF Analysis](https://snorkel.readthedocs.io/en/v0.9.7/packages/_autosummary/labeling/snorkel.labeling.LFAnalysis.html)).

**label_coverage(Y=Optional[List[Doc])**: For each NER Tag, compute the fraction of data points with at least one label (e.g., 60% of data points have PER, 50% have DATE, 40% have LOC). Unlike a classification problem like Snorkel, coverage metrics won't be very helpful without priors on how frequent these entities should appear. For example, 60% of the data points having PER labels (according to LFs) might actually be a good thing - if only 60% of the dataset have PER entities. Accordingly, I'm thinking we should give users the ability to provide the same set of documents with gold labeled spans (if they have such gold label data) so that they can see how much coverage is afforded by their LFs. **Wdyt?**

----------

**label_conflict()**: For each NER Tag, compute the fraction of data points with conflicting (non-abstain labels). There are two broad definitions of conflict:

1. *Disagreement among LFs for shared Label*: If we have 2 LFs for a DATE label (as an example), if they LF1 properly labels a span "June 29" as a date while LF2 does not label "June 29" as a date, we could consider this a conflict. 
2. **Disagreements among LFs for different Labels**: If we have 1 LF for PER and 1 LF for DATE (as an example), if LF1 incorrectly labels a span "June 29" as a date while LF2 correctly labels "June 29" as a date, we could consider this as a conflict. 

These conflicts seem fundamentally different; and given that these analysis tool is meant to help folks debug and improve their LFs, I'm thinking we may want to split `label_conflict()` into two separate functions `shared_label_conflict()`and `diff_label_conflict()`.  **Wdyt?**

Additionally, it seems like this metric would be most useful over spans rather than datapoints, given that we are solving a sequence labeling task rather than classification task. **Wdyt?**

----------

**label_overlap()**: For each NER Tag, compute the fraction of data points with at least two (non-abstain labels). 

Similar to  `label_conflict()`it seems like this metric would be most useful over spans rather than datapoints. **Wdyt?**

**Comment by plison** [comment]
Yes, your suggestions do make sense! Indeed, it's indeed much more informative to get per-label metrics, since most labelling functions are tailored to recognise few specific labels (and ignore the rest). 

Regarding the coverage, it's indeed a difficult problem for sequence labelling  (in contrast to standard classification where each points belong to one class) -- when an LF outputs an "O", we don't know if that means the LF abstains from giving a prediction, or whether the LF predicts that the token does not belong to a category. Yes, one solution is indeed to provide the users the ability to give gold standard docs, such that one could compute the coverage based on them (in that case, one can directly compute the recall, precision and F1 for each label).  On the other hand, this would basically amount to doing a classical evaluation based on test data, and is a quite different kind of exercise than what is typically meant with LF analysis, where the aim is to compare LFs among themselves.

One alternative would be to say that, for a given label (say PER), we look at all the tokens where at least one LF has predicted PER, and we define the coverage in relation to that set. Of course, this estimate of coverage will tend to underestimate the actual coverage (since the denominator will also include false positive tokens for that label), but at least it could be computed without access to gold standard data. Would that make sense for you?

There are also two other aspects that should be ironed out:
- should the metrics be computed at the token-level or entity-level? In other words, should the coverage of e.g. PER entities look at the proportion of PER tokens that are covered, or the proportion of PER entities (that may span multiple words)? I think the easiest would be to provide all metrics at the token-level, but that can of course be discussed. One advantage of computing the metrics at the token-level is that the metrics can be applied to all types of sequence labelling tasks, not just NER.
- when computing the coverage, conflict, and overlap, how should we handle BIO tags? In other words, if a LF outputs B-PER and another LF outputs I-PER, should we count this as a match or as a conflict? It might be worth including an option to allow the user to perform either "strict" matching (i.e. counting B-PER and I-PER as a conflict) or robust matching (i.e. only looking at the category, which means B-PER and I-PER would become a match). 

Thanks again for this great work :-)

**Comment by schopra8** [comment]
Agreed that we need to iron out those cases :) 

1. **Should the metrics be computed at the token-level or entity-level?** Agreed, token-level seems to be a more encompassing approach. Let's go with token level for now ... in the future it's always possible to add bells and whistles to use spans instead.
2. **When computing the coverage, conflict, and overlap, how should we handle BIO tags?** Agreed, we'll have to handle both "strict" and "robust" matching. I'm thinking to prioritize "robust" up-front for the v1, wdyt?

------
On the topic of coverage specifically, can you clarify what the formula would be for coverage in your definition? Reading your description, it seems like this would provide a coverage % for each label function (which is definitely something we should provide, similar to LFAnalysis in Snorkel) -- but it's not clear to me how you would compute a single coverage metric that would encapsulate the coverage provided by combining all of your label functions.

**Comment by plison** [comment]
Ah, my bad, I didn't see the difference between `lf_coverages` and `label_coverages`! Yes, given that we do not know how many tokens should have a non-O label in the first place, I don't quite see how we can provide an implementation for `label_coverage` as in Snorkel. The easiest is indeed to simply drop that function.

I noticed I also did not respond to your question regarding the definition of `label_conflict`. If I understand your description correctly, for `shared_target` you also want to count as conflict cases where one LF predicts e.g. DATE and the other one predicts O, assuming the two LFs are able to predicts DATE? I think it may quickly become a bit message, given that LFs may have very different recalls. I would think it would be easier to simply count as conflict cases where two LFs predicts incompatible, non-O labels. Or where you thinking of doing something different?

**Comment by schopra8** [comment]
Agreed that it would be messy -- since we can't differentiate between Abstain and actively predicting that a token does not belong to a specific entity.

Awesome so to summarize:
- Token-level LF analysis metrics (later introducing possible entity-level metrics)
- "Robust" matching (e.g., I-PER and B-PER are considered the same) for v1 (later introducing "strict" matching)
- Don't implement `label_coverages`
- Implement `lf_coverages` with respect to the set of tokens assigned a particular label (e.g. PER) but 1+ LFs
- `label_conflict` should be implemented with respect to LFs returning different non-null labels for the same tokens (e.g., LF1 labels "Apple" as a ORG while LF2 labels "Apple" as PER)

Does this seem correct? Just want to confirm :)

**Comment by schopra8** [comment]
What's the idiomatic way to get labels for individual tokens, after applying a LF to a Doc object? For spans, I know one can do `doc.spans['lf_name']` but was unsure about how to do so at the token level. Thanks!

**Comment by plison** [comment]
Yes, your summarisation is correct, that's exactly what I had in mind :-)

Coming to your follow-up questions:
- No, there isn't a direct way to get the labels that can be produced by a given LF, unfortunately (without going into the details, the main problem comes from LFs such as DocumentMajorityAnnotator that rely on the outputs of other LFs). However, given a collection of documents annotated with LFs, it's relatively straightforward to loop on the documents and the spans defined in those documents to find out the labels that are produced by each LF:

```
labels_by_lf = {}
for doc in docs:
    for lf_name, spans in doc.spans.items():
        if lf_name not in labels_by_lf:
            labels_by_lf[lf_name] = set()
        labels_by_lf[lf_name].update(span.label_ for span in spans)
```

- I would simply treat all tokens in the same manner, without any special treatment for "gap tokens" or similar. So the "-" would be treated as I-DATE in your example.
- I see your point, but I fear the word "agreement" may give the wrong impression that it provides some annotator agreement metric. In my view, the idea is that, for a given label such as PER, the number of tokens that are marked as PER by at least one LF (among all LFs) can give us a rough estimate of the actual number of PER tokens in the corpus. Of course this will be an overestimate due to false positives, but if we assume that this estimate is not too far off, then the proportion of tokens marked as PER by a given LF divided by this total number will provide a conservative estimate for the LF's recall (and thus the LFs coverage).

**Comment by schopra8** [comment]
Makes sense on all accounts! Let's stick to the term coverage accordingly.

**Comment by schopra8** [comment]
@plison I'm planning on extending the `spans_to_array` helper function so that it can be used, when supporting robust or strict matching.

Looking at the current implementation, I see that the NB states that the "O" tag is given the value 0. But, the enumeration of labels (as written currently) would also assign the value 0 to the first label in `labels`. Is this intentional? Or is this a small bug?

[Code Linked Here](https://github.com/NorskRegnesentral/skweak/blob/main/skweak/utils.py#L454-L467)

**Comment by schopra8** [comment]
Also, is there a standard code formatter and accompanying config file I should be using to make sure my code formatting matches that of the overall library (e.g., yapf or black)? Just want to make sure I get my ducks in a row before creating a pull request for you to review :)

**Comment by plison** [comment]
No, I haven't used any recommended code formatter (but you're right that we probably should!)

**Comment by schopra8** [comment]
@plison I have opened #15 to introduce the LF Analysis tooling. Would love to get your eyes on it and gather your feedback. Please let me know if you'd prefer to connect for a live call - and I can walk you through it at a high level, if that makes it easier to review.
